### ISSUE TITLE
Z: Value types: Nested field class for aconst_init is not patched correctly in non-SVM AOT compilation

### ISSUE THREAD

**Issue by a7ehuo** [eclipse-openj9/openj9#15424]
Value type bytecode `aconst_init` is represented as `newvalue  jitNewValue` in ILGen and later lowered as `new` in CG. If a field of VT is also a VT, we will generate `newvalue  jitNewValue` for the field instance as well. 

I'm testing [enabling AOT compilation for aconst_init](https://github.com/a7ehuo/openj9/commit/53fa21adeb6621d2043acabcfb05f5bf6d747af6), using the following test code, I noticed the field class `SingleFieldPrimitive`  for `NestedFieldsPrimitive` is not patched correctly on AOT load on Z. 

```java
primitive public class NestedFieldsPrimitive {
    public final SingleFieldPrimitive f1;

    public NestedFieldsPrimitive() {
        this.f1 = new SingleFieldPrimitive();
    }

    public NestedFieldsPrimitive(int val) {
        this.f1 = new SingleFieldPrimitive(val);
    }
}
public class TestAOTAConstInit2 {

   public static NestedFieldsPrimitive _nfp;

   public static void test() {
      _nfp = NestedFieldsPrimitive.default; // aconst_init

   public static void main(String[] args) {
      test();
      test();
   }
}
```

```
n7n       treetop                                                                             [     0x3ff88201770] bci=[-1,0,-] rc=0 vc=0 vn=- li=- udi=- nc=1
n6n         newvalue  jitNewValue[#101  helper Method] [flags 0x400 0x0 ] (Identityless sharedMemory )  [     0x3ff88201720] bci=[-1,0,-] rc=2 vc=0 vn=- li=- udi=- nc=2 flg=0x2000
n4n           loadaddr  SingleFieldPrimitive[#383  Static] [flags 0x18307 0x0 ]               [     0x3ff88201680] bci=[-1,0,-] rc=1 vc=0 vn=- li=- udi=- nc=0
n5n           iconst 0                                                                        [     0x3ff882016d0] bci=[-1,0,-] rc=1 vc=0 vn=- li=- udi=- nc=0
n9n       treetop                                                                             [     0x3ff88201810] bci=[-1,0,-] rc=0 vc=0 vn=- li=- udi=- nc=1
n8n         newvalue  jitNewValue[#101  helper Method] [flags 0x400 0x0 ] (Identityless sharedMemory )  [     0x3ff882017c0] bci=[-1,0,-] rc=2 vc=0 vn=- li=- udi=- nc=2 flg=0x2000
n3n           loadaddr  NestedFieldsPrimitive[#382  Static] [flags 0x18307 0x0 ]              [     0x3ff88201630] bci=[-1,0,-] rc=1 vc=0 vn=- li=- udi=- nc=0
n6n           ==>newvalue
```

In AOT compilation, I'm guessing `0x3ff882e0ec0` loads class `SingleFieldPrimitive`. On AOT load, it's incorrectly patched as `TestAOTAConstInit2`. 

**---- AOT Compilation ----**
```
------------------------------
 n7n      (  0)  treetop                                                                              [     0x3ff88201770] bci=[-1,0,-] rc=0 vc=80 vn=- li=2 udi=- nc=1
 n6n      (  3)    new  jitNewValue[#101  helper Method] [flags 0x400 0x0 ] (in &GPR_0034) (highWordZero Unsigned X!=0 skipZeroInit allocationCanBeRemoved sharedMemory )  [     0x3ff88201720] bci=[-1,0,-] rc=3 vc=80 vn=- li=2 udi=- nc=1 flg=0xc004
 n4n      (  0)      loadaddr  SingleFieldPrimitive[#383  Static] [flags 0x18307 0x0 ] (in GPR_0032)  [     0x3ff88201680] bci=[-1,0,-] rc=0 vc=80 vn=- li=2 udi=- nc=0
------------------------------

 [     0x3ff882e0ec0]	                       LGRL    GPR_0032, 0x000003FF882E0D30	
 [     0x3ff882e1190]	                       LGFI    GPR_0033,16	
 [     0x3ff882e1360]	                       AG      GPR_0033,#389 96(GPR13)
 [     0x3ff882e14f0]	                       CLG     GPR_0033,#390 104(GPR13)
 [     0x3ff882ef7a0]	                       Label L0022:	# (Start of internal control flow)	
 [     0x3ff882e15c0]	                       BRC     BL(0x2), Outlined Label L0016	
 [     0x3ff882e1760]	                       LG      &GPR_0034,#391 96(GPR13)
 [     0x3ff882e18f0]	                       STG     GPR_0033,#392 96(GPR13)
 [     0x3ff882ef5e0]	                       ST      GPR_0032,#393 0(&GPR_0034)


     0x3ff6af0009c 0000001c [     0x3ff882e0670] 17 00                                             XR      GPR0,GPR0	
     0x3ff6af0009e 0000001e [     0x3ff882e0ec0] c4 08 00 00 00 a1                                 LGRL    GPR0, 0x000003FF882E0D30, targetAddr=0x000003FF6AF001E0 (offset=0x0000000000000142)	
     0x3ff6af000a4 00000024 [     0x3ff882e1190] c0 11 00 00 00 10                                 LGFI    GPR1,16	
     0x3ff6af000aa 0000002a [     0x3ff882e1360] e3 10 d0 60 00 08                                 AG      GPR1,#389 96(GPR13)
     0x3ff6af000b0 00000030 [     0x3ff882e14f0] e3 10 d0 68 00 21                                 CLG     GPR1,#390 104(GPR13)
```

**---- AOT Load ----**
```
0x3ffdde7f09c {TestAOTAConstInit2.test} +28                       1700         xr        %r0, %r0
0x3ffdde7f09e {TestAOTAConstInit2.test} +30                       c408000000a1 lgrl      %r0, 0x3ffdde7f1e0 J9Class - {TestAOTAConstInit2} //<---- wrong class!! It should be SingleFieldPrimitive
0x3ffdde7f0a4 {TestAOTAConstInit2.test} +36                       c01100000010 lgfi      %r1, 0x10 
0x3ffdde7f0aa {TestAOTAConstInit2.test} +42                       e310d0600008 ag        %r1, 0x60(%r13)  J9VMThread.heapAlloc
0x3ffdde7f0b0 {TestAOTAConstInit2.test} +48                       e310d0680021 clg       %r1, 0x68(%r13)  J9VMThread.heapTop


(kca) what (0x3ffdde7f1e0)
0x3ffdde7f1e0: 0x0000000000240a00 J9Class - {TestAOTAConstInit2}
```
[0623_NestedFieldsPrimitive_AOT.zip](https://github.com/eclipse-openj9/openj9/files/8971220/0623_NestedFieldsPrimitive_AOT.zip)

**Comment by a7ehuo** [comment]
There is a similar issue on X86 but I'm not sure if Z is the same root cause.

For X86-64, it creates a relocation record TR_ClassAddress for the romClass of SingleFieldPrimitive in [omr/compiler/x/amd64/codegen/OMRMemoryReference.cpp](https://github.com/eclipse-openj9/openj9-omr/blob/a57ad601d6664d97958c59884c4d34a0ce2f960a/compiler/x/amd64/codegen/OMRMemoryReference.cpp#L450-L453) for the class load instruction. The TR_ClassAddress relocation record requires cpIndex for the class symref, but when the field of VT is VT, the cpIndex for the field class SingleFieldPrimitive is unknown, so [it is set as 0 in ILGen](https://github.com/a7ehuo/openj9/blob/53fa21adeb6621d2043acabcfb05f5bf6d747af6/runtime/compiler/ilgen/Walker.cpp#L6744). Because cpindex is incorrect, the patched address is also incorrect.

**Comment by r30shah** [comment]
Posting a link of similar code for Z that @a7ehuo has referenced from x86 in previous commit in [1] and [2]. We create similar relocation record. Relocating the ClassAddress in non-SVM AOT would be similar on Z and x86 I think. May be @dsouzai / @jdmpapin  can share more light on what will be needed to get `SingleFieldPrimitive` relocated. 

[1]. https://github.com/eclipse/omr/blob/a57ad601d6664d97958c59884c4d34a0ce2f960a/compiler/z/codegen/ConstantDataSnippet.cpp#L93-L120
[2]. https://github.com/eclipse-openj9/openj9/blob/7b717f70683c05a5304f8861e2ee675380753e65/runtime/compiler/z/codegen/S390AOTRelocation.cpp#L53-L73

**Comment by a7ehuo** [comment]
Thanks @r30shah for the above information! Now I have a better understanding of it on Z. I think Z has the same issue as X86 [1]. One potential workaround we're considering is to abort AOT compilation if it's non-SVM AOT and if the VT has a field that is also a VT. Since in this case, [cpindex for the field is unknown in ILGen](https://github.com/eclipse-openj9/openj9/blob/da77f9f134bd46b7a5e0a608101f4d55eb4da8c9/runtime/compiler/ilgen/Walker.cpp#L6720-L6725). 


[1] 
The instruction `0x3ff6af0009e`  is `loadaddr  SingleFieldPrimitive`. It has target address `0x000003FF6AF001E0` which is a data snippet. It has a relocation record `TR_ClassAddress` with PC offset `0160` (or offset `01d0`). On AOT load, ` (offset 1d0)` is patched as `TestAOTAConstInit2` because cpindex is 0. 

```
n4n      (  0)      loadaddr  SingleFieldPrimitive[#383  Static] [flags 0x18307 0x0 ] (in GPR_0032)  [     0x3ff88201680] bci=[-1,0,-] rc=0 vc=80 vn=- li=2 udi=- nc=0
...
 [     0x3ff882e0ec0]	                       LGRL    GPR_0032, 0x000003FF882E0D30	
...
     0x3ff6af0009e 0000001e [     0x3ff882e0ec0] c4 08 00 00 00 a1                                 LGRL    GPR0, 0x000003FF882E0D30, targetAddr=0x000003FF6AF001E0 (offset=0x0000000000000142)	
```

```
TR_ClassAddress (40)                /root/home/ahuo/src/openj9-openjdk-jdk.valuetypes/omr/compiler/z/codegen/ConstantDataSnippet.cpp   114      01d0       0160 000003FF88201680
TargetAddress1:000003FF8824D3C0,  TargetAddress2:FFFFFFFFFFFFFFFF
...
        6a5ff14c  34   TR_ClassAddress (40)           2     Abs       01d0 
 Address Relocation (TR_ClassAddress): inlinedIndex = -1, constantPool = 000003FF496D0868, CPI = 0, flags = 0

```

```
000003FF6AF001E0 00000160                                                 Snippet Label L0018:		; Constant Data Snippet
     0x3ff6af001e0 00000160                      00 00 03 ff 49 6d 43 e8    DC   	0x00000000007fde00
```

```
TR_ClassAddress 000003FFDDE7C92C
	size 22 type 40 flags 0 reloFlags 0
	inlined site index FFFFFFFFFFFFFFFF
	constant pool 000003FF496D0868
	cpIndex 0000000000000000. //<------ invalid CPIndex
	reloLocation: from 000003FFDDE7C94C at 000003FFDDE7F1E0 (offset 1d0)
		computeNewConstantPool: newCP 0000000000240BC0
	computeNewClassObject: resolvedClass 0000000000240A00
	computeNewClassObject: resolvedClassName TestAOTAConstInit2 //<------ patched wrong class 
```

**Comment by jdmpapin** [comment]
Aborting non-SVM AOT compilation when this situation arises is the most expedient fix. I don't believe there is currently any good way for non-SVM AOT to relocate the class in this situation. At least there's no general way. An exception could be made for a class defined by the bootstrap class loader (in which case the `TR_ClassAddress` will automatically be changed into a `TR_ArbitraryClassAddress`). We could also scrounge around for a constant pool entry that happens to have resolved to the needed class, and indicate that constant pool entry when it's available, but I really don't think we should go down that road

**Comment by jdmpapin** [comment]
It's also possible to try to make `TR_ArbitraryClassAddress` more general, but IMO any such effort would be better spent toward requiring SVM for all AOT compilations

**Comment by a7ehuo** [comment]
Thanks for the clarification @jdmpapin! I think I'll go with aborting AOT compilation in ILGen if it's non-SVM AOT + the VT has a field that is also a VT.  In inlining allocation, different types of relocation record are created in different platforms: `TR_ClassAddress` on X86 and Z; `TR_ConstantPool` on Power. It wouldn't be trivial to make `TR_ArbitraryClassAddress` work on all platforms when the issue only exists in this specific case. 

>The usual value indicating an unknown or nonexistent constant pool index is -1

Thanks for pointing it out. I'll make sure to pass `-1` when CPIndex is unknown in this case.

**Comment by jdmpapin** [comment]
It's redundant, but it should be fine. For these accesses generated by ilgen, the class came from the constant pool, and the class from constant pool validation (i.e. the first validation) will check on load that the constant pool entry is resolved to a class with the expected layout\*

The arbitrary class validation should usually find the same class. It's possible for it to fail to find a loader or find the wrong loader, and then as a result fail to find the class or (I think) even find a different class. If it fails to find a class, or if it finds a class with a different layout,\* then I'm pretty sure the load will just fail. This would be a spurious failure, since we're only really relying on the class as found from the constant pool. If it finds a different class that happens to also match the expected layout, then non-SVM AOT won't notice the discrepancy, but again it's irrelevant to the offsets for these ilgen-generated loads

There are some (pretty contrived but) possible scenarios in which a non-SVM AOT compilation could get hold of a single class from multiple sources, reason about that class during compilation in a way that implicitly assumes that those sources will all agree at load time, only for them to actually disagree at load time, resulting in a discrepancy of the kind that I mentioned above. I don't know of a straightforward way to prevent such an issue in general in non-SVM AOT - or rather, SVM is the straightforward way to prevent it - but I don't think that such problems are necessarily any worse for value types than for general classes

@dsouzai, anything to add or correct?

\*The fine print: Really, the validations check that the class chain matches. With flattening, if we haven't already, we'll have to make sure that the class chain also specifies which fields are flattened and includes the class chains of (transitively) all of the types of those fields

**Comment by dsouzai** [comment]
>@dsouzai, anything to add or correct?

Yeah, just regarding:

>*The fine print: Really, the validations check that the class chain matches. With flattening, if we haven't already, we'll have to make sure that the class chain also specifies which fields are flattened and includes the class chains of (transitively) all of the types of those fields

We haven't added the notion of flattened fields in the class chain yet, so that's something we need to think about and design. I had spent some time thinking about this when this subtlety was first brought up a year or more ago, and I have a vague recollection of a relatively efficient way of handling the class chain validation of a flattened type but I need to see if I can dig up what that idea was.

**Comment by a7ehuo** [comment]
Fixed in #15460
